### PR TITLE
Shared VPC support

### PR DESCRIPTION
**Pull request by atm**
Adds support for shared VPCs, as well as explicitly sets the default service account.

### CODE REVIEW

File: provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeInstanceTemplateConfigurationProperty.java [hunk 1]
```diff
@@ -63,6 +63,33 @@
       .required(false)
       .build()),
 
+  NETWORK_PROJECT(new SimpleConfigurationPropertyBuilder()
+      .configKey("networkProject")
+      .name("Network Project")
+      .defaultDescription(
+          "The project the nework belongs to.<br />" +
```
Comment (bhavanki) [inline provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeInstanceTemplateConfigurationProperty.java]: nit: network (spelling)

File: provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeInstanceTemplateConfigurationProperty.java [hunk 2]
```diff
@@ -63,6 +63,33 @@
...
+              "<a target='_blank' href='https://cloud.google.com/compute/docs/networking#networks'>More Information</a>")
+      .defaultValue(null)
+      .required(false)
+      .build()),
+
+  ASSIGN_EXTERNAL_IPS(new SimpleConfigurationPropertyBuilder()
+      .configKey("assignExternalIPs")
+      .name("Assign External IPs")
+      .defaultDescription("Assign external IP addresses to created instances.  Cloudera director will" +
```
Comment (bhavanki) [inline provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeInstanceTemplateConfigurationProperty.java]: Cloudera Altus Director

Also, s/will still require/may require - it's possible to point Director to package repository mirrors within a private network, when outbound access is prohibited.

File: provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeInstanceTemplateConfigurationValidator.java [hunk 3]
```diff
@@ -475,7 +476,7 @@ void checkNetwork(Configured configuration,
 
     String networkName = configuration.getConfigurationValue(NETWORK_NAME, localizationContext);
 
-    if (networkName != null) {
+    if (networkName != null && !networkName.equals("default")) {
```
Comment (bhavanki) [inline provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeInstanceTemplateConfigurationValidator.java]: Why is this check here? There is special handling below as well in GoogleComputeProvider. My guess is that the assumption is that the default network will alwaus be there, so why spend time validating it or creating a network interface specifically naming it. However, it looks like an organization can opt out of one existing:

https://cloud.google.com/vpc/docs/vpc#default-network

So I think it's better to remove the special handling, but anyone with more knowledge can correct me here.

File: provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeProvider.java [hunk 4]
```diff
@@ -70,6 +73,8 @@
 import com.google.api.services.compute.model.Operation;
 import com.google.api.services.compute.model.Scheduling;
 import com.google.api.services.compute.model.Tags;
+import com.google.api.services.compute.model.ServiceAccount;
```
Comment (bhavanki) [inline provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeProvider.java]: super nit! Move this above the import of Tags, for alphabetical order

(and below, move Collections up too)

File: provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeProvider.java [hunk 5]
```diff
@@ -355,13 +376,34 @@ public void allocate(GoogleComputeInstanceTemplate template,
...
       String applicationNameVersionTag = Names.buildApplicationNameVersionTag(applicationProperties);
       // Massage it into a form acceptable for use as a tag (only allows lowercase letters, numbers and hyphens).
       applicationNameVersionTag = applicationNameVersionTag.toLowerCase().replaceAll("\\.|/", "-");
-      tags.setItems(Lists.newArrayList(applicationNameVersionTag));
+
+      ArrayList<String> allTags = Lists.newArrayList();
+
+      String userTags = template.getConfigurationValue(INSTANCE_TAGS, templateLocalizationContext);
+      if (userTags != null) {
+        //  TODO I can't figure out a way to get a list out of the template for these properties
```
Comment (bhavanki) [inline provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeProvider.java]: This is totally fine 😄

File: provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeProvider.java [hunk 6]
```diff
@@ -294,24 +300,39 @@ public void allocate(GoogleComputeInstanceTemplate template,
...
 
       if (subnetName != null) {
         String region = getConfigurationValue(GoogleComputeProviderConfigurationProperty.REGION,
             providerLocalizationContext);
-        networkInterface.setSubnetwork(Urls.buildSubnetUrl(projectId, region, subnetName));
+        String subnetwork = Urls.buildSubnetUrl(networkProject, region, subnetName);
+        networkInterface.setSubnetwork(subnetwork);
       }
+      //  only specify network if the subnetwork isn't set
+      else if (!networkName.equals("default")) {
```
Comment (meowfaceman) [inline provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeProvider.java]: No null check necessary here?

**Comment by atm** [comment]
Thanks for the reviews, @bhavanki and @meowfaceman. I just pushed two commits that should address them all.

@bhavanki - with regard to the feedback regarding the commit message, is that on you to do when merging this pull request? Or would you like me to squash all these commits into a single commit with the commit message you suggested? Or something else?

File: provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeProvider.java [hunk 7]
```diff
@@ -294,24 +300,39 @@ public void allocate(GoogleComputeInstanceTemplate template,
...
 
       if (subnetName != null) {
         String region = getConfigurationValue(GoogleComputeProviderConfigurationProperty.REGION,
             providerLocalizationContext);
-        networkInterface.setSubnetwork(Urls.buildSubnetUrl(projectId, region, subnetName));
+        String subnetwork = Urls.buildSubnetUrl(networkProject, region, subnetName);
+        networkInterface.setSubnetwork(subnetwork);
       }
+      //  only specify network if the subnetwork isn't set
+      else if (networkName != null && !networkName.equals("default")) {
```
Comment (bhavanki) [inline provider/src/main/java/com/cloudera/director/google/compute/GoogleComputeProvider.java]: Here's another spot where the "default" group is handled specially, and probably shouldn't be.

**Comment by bhavanki** [comment]
> Thanks for the reviews, @bhavanki and @meowfaceman. I just pushed two commits that should address them all.
> 
> @bhavanki - with regard to the feedback regarding the commit message, is that on you to do when merging this pull request? Or would you like me to squash all these commits into a single commit with the commit message you suggested? Or something else?

If you can squash, that would be great, just to have one commit in the history. I actually don't know how github decides what the commit message(s) should be, but I'll figure that out and let y'all know what to do about including bpodgursky in it.

**Comment by bpodgursky** [comment]
Thanks guys for cleaning my changes up and getting them merged : )

**Comment by bhavanki** [comment]
Thanks @bpodgursky !

**Pull request merged**